cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/runtimejob/DataprocRuntimeJobManager.java

```diff
@@ -320,26 +327,56 @@ private List<LocalFile> getRuntimeLocalFiles(Collection<? extends LocalFile> run
   private LocalFile uploadFile(String bucket, String targetFilePath,
                                LocalFile localFile) throws IOException, StorageException {
     BlobId blobId = BlobId.of(bucket, targetFilePath);
-    BlobInfo blobInfo = BlobInfo.newBuilder(blobId).setContentType("application/octet-stream").build();
-
+    String contentType = "application/octet-stream";
+    BlobInfo blobInfo = BlobInfo.newBuilder(blobId).setContentType(contentType).build();
+    Storage storage = getStorageClient();
+    boolean preConditionFailure = false;
 
     LOG.debug("Uploading a file of size {} bytes from {} to gs://{}/{} ", localFile.getSize(), localFile.getURI(),
               bucket, targetFilePath);
-    Bucket bucketObj = getStorageClient().get(bucket);
+    Bucket bucketObj = storage.get(bucket);
     if (bucketObj != null) {
       LOG.debug("File's Location type : {} and Location : {}. ", bucketObj.getLocationType(), bucketObj.getLocation());
     }
-    try (InputStream inputStream = openStream(localFile.getURI());
-         WriteChannel writer = getStorageClient().writer(blobInfo)) {
-      ByteStreams.copy(inputStream, Channels.newOutputStream(writer));
+    try {
+      uploadFileUtil(localFile.getURI(), storage, blobInfo, Storage.BlobWriteOption.doesNotExist());
+    } catch (StorageException e) {
+      if (e.getCode() == HttpURLConnection.HTTP_PRECON_FAILED) {
+        // Precondition fails means the blob already exists, most likely happens due to retries
+        // https://cloud.google.com/storage/docs/request-preconditions#special-case
+        // Overwrite the file
```

Comment:
Do we want to add a log statement here to let the user know that a file was overwritten?

Reply:
Yeah, log is added on line [#357](https://github.com/cdapio/cdap/pull/14216/files#diff-b7675d51d200ad4368d96b266bc04c96d4ecd20e5d1e1a6510f43fee970ffacbR357)